Add quickstart guide for ohttp-relay on basic ubuntu server

benalleng opened the pull request:
This issue https://github.com/payjoin/ohttp-relay/issues/37 outlines a need for some documentation for a quickstart guide on a docker server with an nginx proxy.

You can take a look at the pretty markdown here https://github.com/benalleng/payjoin.org/blob/ohttp-relay-tutorial/docs/tutorials/quickstart-ohttp-relay.md

Still a WIP as there is still some text i need to cleanup and I have some outstanding questions

A few oustanding questions remain with this guide.
 - ~~Figure out why my current implementation is not quite working correctly~~ I think this is ready for some double checks as I am able to proxy to the payjo.in directory successfully
 - What are some basic test vectors, whether directing them to rust-payjoin to setup the relay in the tests somehow or just some very basic curl requests to make sure their server is receiving well
     - I think this `curl -vk --proxy-insecure --proxy https://{you-public-ip} https://payjo.in/ohttp-keys --output -` command does an ok job of ensuring that the proxy successfully connects and is able to then receive data from the payjo.in directory
 - ~~Should we strip out some of the details like PORT selection in our `.conf` files and leave it as they exist [here](https://github.com/payjoin/ohttp-relay/blob/main/nginx.conf.template)~~ A quickstart guide like this is the time to be opinionated

benalleng committed:
Add quickstart guide for ohttp-relay on basic ubuntu server

This issue https://github.com/payjoin/ohttp-relay/issues/37 outlines
a need for some documentation for a quickstart guide on a docker server
with a nginx proxy.

This iteration uses the existing dockerfile with a nginx reverse proxy
in front.

DanGould commented on docs/tutorials/quickstart-ohttp-relay.md:
```diff
@@ -0,0 +1,92 @@
+# Quickstart ohttp relay
+
+We'll demostrate how you can setup an ohttp server on common cloud insfrastructure on something like [](483) using the `ohttp-relay` crate in Payjoin Dev Kit. This should take about 30 minutes. This tutorial assumes you have a basic understanding of the AWS infrastructure.
```
empty link. What is "something like [](483)" referencing?

Not using AWS afaict

benalleng replied:
Yeah, sorry the copy still needs some cleanup. I had started making the guide thinking I would need to make it aws specific but its pretty easy to see that's not necessary.

DanGould commented on docs/tutorials/quickstart-ohttp-relay.md:
```diff
@@ -0,0 +1,92 @@
+
+We'll demostrate how you can setup an ohttp server on common cloud insfrastructure on something like [](483) using the `ohttp-relay` crate in Payjoin Dev Kit. This should take about 30 minutes. This tutorial assumes you have a basic understanding of the AWS infrastructure.
+
+The network parameters could be changed to support any Bitcoin network, but this tutorial will use the Bitcoin Signet network. We will ignore most errors and Payjoin Session persistence for simplicity. Reference the [`payjoin-cli` source](https://github.com/payjoin/rust-payjoin/tree/master/payjoin-cli) reference implementation for more information on those implementation details.
```
Do we even need to reference a bitcoin network to get this working?

DanGould commented on docs/tutorials/quickstart-ohttp-relay.md:
```diff
@@ -0,0 +1,92 @@
+
+First, you must install docker and nginx on your server.
+
+Install docker ond nginx on a fresh Ubuntu server:
```
ond/and

DanGould commented on docs/tutorials/quickstart-ohttp-relay.md:
````diff
@@ -0,0 +1,92 @@
+
+```sh 
+#/etc/nginx/nginx.conf
+load_modules /usr/lib/nginx/modules/ngx_mod_stream.so;
````
I think we need instructions to get the nginx binary with ngx_mod_stream support. I don't believe it's compiled in by default.

benalleng replied:
The simplest way to approach this would be I think to add `libnginx-mod-stream` to our install step

DanGould reviewed (commented):
> just some very basic curl requests to make sure their server is receiving well

This is the quickest way for us to meaningfully get off zero. I like it.

> Should we strip out some of the details like PORT selection in our .conf files and leave it as they exist [here](https://github.com/payjoin/ohttp-relay/blob/main/nginx.conf.template)

A concrete tutorial can state its opinion on explicit port definition IMO.

benalleng commented on docs/tutorials/quickstart-ohttp-relay.md:
````diff
@@ -0,0 +1,103 @@
+Check your work by running the curl request below to do a quick check to make sure your server is receiving well.
+
+For testing we will include th `-vk and --proxy-insecure` flags to ensure we can get past any warnings or errors our curl request might send us with invalid cert signatures. For final checks in prod we recommend removing these flags to ensure your keys and certs are working and up-to-date.
+
+This curl request should occur in 2 stages
+1. The proxy CONNECT request that passes through your relay to the `https://payjo.in` directory
+2. The GET request on `/ohttp-keys` that will return a binary encoded output
+A successful test should return a 200/OK response on both of these steps
+
+```sh
+curl -vk --proxy-insecure --proxy https://{your-relay-public-ip} https://payjo.in/ohttp-keys --output -
+```
````
I assume that these keys could change is there a nice way to check and make sure the keys we are receiving are what we expect?

benalleng committed:
Update the docs with some more optimizations and tips

This commit adds some helpful commands and tips to make it easier to
setup and as simple reminders about what is needed for the relay to be
setup correctly.

benalleng commented on docs/tutorials/quickstart-ohttp-relay.md:
````diff
@@ -0,0 +1,152 @@
+We can use some built-in flags in docker to run our ohttp-relay in the background.
+
+```sh
+sudo docker run -d --restart unless-stopped --name ohttp-relay -p 3000:3000 \
+    -e PORT=3000 \
+    -e GATEWAY_ORIGIN=https://payjo.in \
+    ohttp-relay
````
This can be updated for the latest payjoin-mailroom with a more simplified setup relying more on the config.toml.
```suggestion
sudo docker run -d --restart unless-stopped \
  --name payjoin-mailroom \
  -p 3000:3000 \
  -v ~/payjoin-mailroom/config.toml:/config.toml:ro \
  -v ~/payjoin-mailroom/data:/data \
  payjoin/payjoin-mailroom:payjoin-mailroom-0.1.1 \
  --config /config.toml
```
